### DISCUSSION TITLE
Code style guide

### DISCUSSION THREAD

**Discussion by Derkades** [NamelessMC/Nameless#1786] [General]
I'll edit this post with any conclusions drawn in replies.

For HTML:
- 2 spaces indentation
- XHTML syntax

For php, default phpcs settings with the following differences:

- Use keywords like functions (e.g. `echo("Something")`)
- 120 character line limit instead of 85
- `{` on the same line for classes and functions

TODO: Figure out how to create phpcs.xml configuration file with these settings

How to use on vscode on ubuntu:
```
apt install php-fpm php-codesniffer
```
Install `phpcs` extension in vscode

For easy documentation, install the `phpdoc` extension to automatically generate phpdoc comments with `ctrl`+`shift`+`i`

**Comment by Derkades** [comment]
Most important right now: tabs or spaces? how many spaces? possibly different number of spaces for HTML and PHP? Should code in `<?php` blocks be indented? How about css in `<style>`?

**Comment by samerton** [comment]
I'd say 2 space indentations, like so:

```php
<!doctype html>
<html>
  <head>
    <title>Test</title>
    <style>
      body {
        background-color: #000;
        color: #fff;
      }
    </style>
  </head>

  <body>
    <?php
    foreach ($var as $item) {
      echo $item;
    }
    ?>
  </body>
</html>
```

**Comment by Derkades** [comment]
phpcs suggests no parentheses for statements (so `require 'file.php'` not `require('file.php')`, what do you think?

**Comment by partydragen** [comment]
well require('file.php') looks nicer xD

**Comment by samerton** [comment]
Agree with partydragen, for me require() looks nicer

RE https://github.com/NamelessMC/Nameless/commit/7e0e98862a6fa95a2219313a83834df06790bca6#diff-59736aecc611cda3c8a10fbb4681b15045c56f8c2a786bd3c154dacdcb03abbb

Because HTML can end up with some deep indents, can we keep HTML as 2 space indentations, also indent <head> after <html>

**Comment by Derkades** [comment]
> Because HTML can end up with some deep indents, can we keep HTML as 2 space indentations

I would do this for HTML templates that are separated from PHP code, but in this case one file had both and IDEs only support one indentation type per file. It can be done manually if you really want to though.

I haven't touched any "real" php / template files yet, I am waiting for party's PR.

**Comment by Derkades** [comment]
For HTML, use XHTML or HTML syntax? (`<br />` or `<br>`, `<img />` or `<img>`, etc). There doesn't seem to be an objectively correct answer to this, HTML5 allows both.

**Reply by samerton** [reply]
XHTML would be preferred here as it's more "proper" 😄 

So `<br />`, `<img />` etc -  all tags should be correctly closed

**Reply by Derkades** [reply]
I wouldn't say it's more "proper", the HTML5 specification says both are appropriate and does not recommend one over the other. In fact, the HTML5 specification uses HTML4 (not XHTML) syntax in all its examples. 

It really doesn't matter though and if you like looking at `<br />` better that's what we'll do :+1:

**Reply by samerton** [reply]
Maybe I'm just old school but the thought of unclosed tags really feels odd to me

**Reply by Derkades** [reply]
It does for many people. Like spaces vs tabs there's really not a correct answer (well, except for YAML and some other languages).
